### ISSUE TITLE
Default value for (non-static) options

### ISSUE THREAD

**Issue by xml-project** [xproc/3.0-specification#506]
Currently we say on the [default value](http://spec.xproc.org/master/head/xproc/#p.option) for (non-static) option:

> It must be a statically valid expression at that point. Consequently, if it contains variable references, they can only be references to preceding options on the step.It is a static error (err:XS0071) if the XPath expression refers to the name of a variable that is not present in the in-scope bindings. It is a dynamic error (err:XD0026) if the select expression makes reference to the context node, size, or position.

1. I would expect static options/variables lexically preceding the step signature to be visible too.
2. Since the XPath expression in `@select` has to be examined during static analysis to raise XS0071, I would suggest that making reference to a context node etc. is a static error too.

**Comment by ndw** [comment]
“Maybe.”

Consider

```xml

<p:declare-step …>

  <p:variable name="x" select="1"/>

  <p:declare-step type="ex:foo">
    <p:option name="a" select="$x"/>
    <p:option name="b" select="$x"/>
    …
  </p:declare-step>

  <p:variable name="x" select="2"/>

  <ex:foo>
    <p:with-option name="b" select="$x"/>
    <!-- ??? -->
  </ex:foo>

  …
</p:declare-step>
```

At the point marked `???`, what are the values of $a and $b and why?

**Comment by xml-project** [comment]
Answer:
a=1, b=2

Proof (Assuming for first `p:variable` has a missing `(@static='true')`:

1. "The values of static options and variables are computed during static analysis." [see here](http://spec.xproc.org/master/head/xproc/#statics)
2. We could therefor imagine every occurrence of `$x` to be replaced by the value computed  for `$x`.
3. The declaration `ex:foo` would then be equivalent with:
````
<p:declare-step type="ex:foo">
    <p:option name="a" select="1"/>
    <p:option name="b" select="1"/>
    …
</p:declare-step>
````
4. It does not matter whether the second variable has a missing `@static=true' (At that position it might be static or dynamic).
5. Since no value is provided for option `a` on the call of `ex:foo`, the default value is used. This is "1", so a=1.
6. A value for `b` is provided, and the provided value is "2", so b=2.

Q.E.D.?

**Comment by ndw** [comment]
Okay. I agree that's the answer. From an implementation perspective, it's quite interesting. It means that the binding for `$x` for one expression is different than the binding for `$x` for the other expression.

I implemented bindings like pipes between steps so there's no obvious way to (a) have two different bindings for `$x` or distinguish between them. But I'm game if you are, I guess.

**Comment by xml-project** [comment]
> I implemented bindings like pipes between steps so there's no obvious way to (a) have two different bindings for $x or distinguish between them.

Similar here: The XDM-value is produced and then send to the implementation of the step on which the $x occurred. 
But with the arrival of static variables/option and the requirement, that the values (and errors) have to be evaluated statically, I do not handle them as messages between actors. I use a classical value map for static variables/option. This was why I said, that my implementation broke because of static options/variables.

What about the second point in my initial posting: Raising a static error?

A third point occurred to me in the meantime: What if the `p:option` has an `@as` and the computed value for `@select` does not conform to the given SequenceType. I would opt for this to be a *dynamic error* raised iff the default value is known to be used. Sounds reasonable?

**Comment by ndw** [comment]
I'd over looked the fact that that first variable has to be static. Yes, that helps.

If we're going to evaluate the defaults statically, then yes, raising the error statically makes sense.

I can go either way on the third point above. I don't see any particular advantage to the user to allow them to specify a bogus default value that only sometimes raises an error. But I don't object to delaying the check to runtime.

**Comment by gimsieke** [comment]
Resolution: Static variables may not be shadowed by dynamic variables.

**Comment by ndw** [comment]
Should have been marked fixed by #615
